File: 69.x-的平方根.java
/*
 * @lc app=leetcode.cn id=69 lang=java
 *
 * [69] x 的平方根
 */

// @lc code=start
class Solution {
    public int mySqrt(int x) {
        int l = 1, r = x;
        while (l <= r) {
            int mid = l + (r - l) / 2;
            if (mid > x / mid) {
                r = mid - 1;
            } else if (mid < x / mid) {
                l = mid + 1;
            } else {
                return mid;
            }
        }
        return r;
    }

    // 变形题：小数点精确到 xxx
    public float mySqrt2(int c) {
        double e = 1e-3;    // 精确到 3 位小数。
        double x = c;
        double y = (x + c / x) / 2;
        while (Math.abs(x - y) > e) {
            x = y;
            y = (x + c / x) / 2;
        }
        return (float) x;
    }
}
// @lc code=end

File: 69.x-的平方根.kt
/*
 * @lc app=leetcode.cn id=69 lang=kotlin
 *
 * [69] x 的平方根
 */

// @lc code=start
class Solution {
    /**
     * 方法：
     * 1. 暴力求解
     * 2. Binary search
     * 3. 牛顿迭代法
     */
    fun mySqrt(x: Int): Int {
        var start = 1
        var end = x

        while (start <= end) {
            val mid = start + (end - start) / 2
            if (mid > x / mid) {    // 注意 mid * mid 可能超出 int 的范围。所以可以改成除法：mid > x / mid
                end = mid - 1
            } else if (mid * mid < x) {
                start = mid + 1
            } else {
                return mid
            }
        }

        return end
    }
}
// @lc code=end

